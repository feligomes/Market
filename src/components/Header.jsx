import React, { useContext, useEffect } from "react";
import styled from "@emotion/styled";
import {
  getCartProducts,
  getCartCost,
  getNumberProductsInCart,
} from "./Products/redux/selector";
import { setNumberProductsInCart, setCartCost } from "./Products/redux/slice";
import { useSelector } from "react-redux";
import Badge from "@mui/material/Badge";
import ShoppingCart from "@mui/icons-material/ShoppingCart";
import { CART_SCREEN } from "../shared/constants/screenNames";
import Context from "../context/Context";
import NumberFormat from "react-number-format";
import { MOBILE_MAX_WIDTH } from "../shared/constants/Common";
import { useMediaQuery } from "react-responsive";
import { useDispatch } from "react-redux";
import Tooltip from "@mui/material/Tooltip";

const HeaderSection = styled.div`
  display: flex;
  flex-direction: row;
  align-items: center;
  padding: 0px 60px;
  justify-content: space-between;
  background: #1a1734;
  height: 60px;
  color: white;
  font-weight: bold;
  top: 0px;
  z-index: 100;
  position: sticky;
`;

const CartSection = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  cursor: pointer;
`;

const Header = () => {
  const cartProducts = useSelector(getCartProducts);
  const dispatch = useDispatch();

  const numberProductsInCart = useSelector(getNumberProductsInCart);
  const cartCost = useSelector(getCartCost);

  const isMobile = useMediaQuery({ maxWidth: MOBILE_MAX_WIDTH });
  const { redirect } = useContext(Context);

  useEffect(() => {
    if (cartProducts) {
      let totalProducts = 0;
      let totalCost = 0;
      cartProducts.forEach((product) => {
        totalProducts += product.quantity;
        totalCost += product.quantity * product.price;
      });
      dispatch(setNumberProductsInCart(totalProducts));
      dispatch(setCartCost(totalCost));
    }
  }, [cartProducts]);

  const handleClick = () => {
    if (numberProductsInCart > 0) {
      redirect(CART_SCREEN);
    }
  };

  return (
    <HeaderSection>
      <span style={{ fontSize: isMobile ? "15px" : "22px" }}>
        Morsum Market
      </span>
      <Tooltip
        title={
          numberProductsInCart === 0 ? "Add items to the cart to continue" : ""
        }
      >
        <CartSection
          onClick={() => {
            handleClick();
          }}
        >
          <span style={{ marginRight: 15 }}>Go to Cart</span>
          <Badge badgeContent={numberProductsInCart} color="primary">
            <ShoppingCart />
          </Badge>
          <NumberFormat
            style={{
              visibility: numberProductsInCart === 0 && "hidden",
              marginLeft: 15,
            }}
            value={cartCost}
            displayType={"text"}
            thousandSeparator={true}
            prefix={"$"}
            decimalScale={2}
          />
        </CartSection>
      </Tooltip>
    </HeaderSection>
  );
};

export default Header;
