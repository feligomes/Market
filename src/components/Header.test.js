import React from "react";
import renderer from "react-test-renderer";
import { Provider } from "../context/Context";
import Header from "./Header";

const cart = [
  {
    id: 1,
    quantity: 3,
    price: 109.95,
    image: "https://fakestoreapi.com/img/81fPKd-2AYL._AC_SL1500_.jpg",
    title: "Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops",
  },
  {
    id: 2,
    quantity: 3,
    price: 22.3,
    image:
      "https://fakestoreapi.com/img/71-3HjGNDUL._AC_SY879._SX._UX._SY._UY_.jpg",
    title: "Mens Casual Premium Slim Fit T-Shirts ",
  },
  {
    id: 3,
    quantity: 2,
    price: 55.99,
    image: "https://fakestoreapi.com/img/71li-ujtlUL._AC_UX679_.jpg",
    title: "Mens Cotton Jacket",
  },
];

jest.mock("react-redux", () => ({
  useSelector: jest.fn().mockImplementation((selector) => selector()),
  useDispatch: jest.fn(),
}));

jest.mock("./Products/redux/selector", () => ({
  getCartProducts: jest.fn().mockReturnValue(cart),
  getCartCost: jest.fn().mockReturnValue(10),
  getNumberProductsInCart: jest.fn().mockReturnValue(10),
}));

it("renders correctly", () => {
  const mockRedirect = (value) => value;
  const mockActualPage = (value) => value;
  const tree = renderer
    .create(
      <Provider value={{ actualPage: mockActualPage, redirect: mockRedirect }}>
        <Header />
      </Provider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
