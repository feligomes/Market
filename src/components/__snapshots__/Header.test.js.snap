// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="css-1s0vyms"
  style={
    Object {
      "padding": "0px 60px",
    }
  }
>
  <span
    style={
      Object {
        "fontSize": "22px",
      }
    }
  >
    Market
  </span>
  <div
    aria-label=""
    aria-labelledby={null}
    className=" css-7fv8xe"
    data-mui-internal-clone-element={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onTouchEnd={[Function]}
    onTouchStart={[Function]}
  >
    <span
      style={
        Object {
          "marginRight": 15,
        }
      }
    >
      Go to Cart
    </span>
    <span
      className="BaseBadge-root MuiBadge-root css-1c32n2y-MuiBadge-root"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="ShoppingCartIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 18c-1.1 0-1.99.9-1.99 2S5.9 22 7 22s2-.9 2-2-.9-2-2-2zM1 2v2h2l3.6 7.59-1.35 2.45c-.16.28-.25.61-.25.96 0 1.1.9 2 2 2h12v-2H7.42c-.14 0-.25-.11-.25-.25l.03-.12.9-1.63h7.45c.75 0 1.41-.41 1.75-1.03l3.58-6.49c.08-.14.12-.31.12-.48 0-.55-.45-1-1-1H5.21l-.94-2H1zm16 16c-1.1 0-1.99.9-1.99 2s.89 2 1.99 2 2-.9 2-2-.9-2-2-2z"
        />
      </svg>
      <span
        className="BaseBadge-badge MuiBadge-badge MuiBadge-standard MuiBadge-invisible MuiBadge-anchorOriginTopRight MuiBadge-anchorOriginTopRightRectangular MuiBadge-overlapRectangular MuiBadge-colorPrimary css-106c1u2-MuiBadge-badge"
      />
    </span>
    <span
      style={
        Object {
          "marginLeft": 15,
          "visibility": false,
        }
      }
    />
  </div>
</div>
`;
